awx/main/models/events.py

```diff
@@ -484,29 +487,45 @@ def _update_host_summary_from_stats(self, hostnames):
             if not self.job or not self.job.inventory:
                 logger.info('Event {} missing job or inventory, host summaries not updated'.format(self.pk))
                 return
-            qs = self.job.inventory.hosts.filter(name__in=hostnames)
             job = self.job
+
+            from awx.main.models import Host, JobHostSummary  # circular import
+            all_hosts = Host.objects.filter(
+                pk__in=self.host_map.values()
+            ).only('id')
+            existing_host_ids = set(h.id for h in all_hosts)
+
+            summaries = dict()
             for host in hostnames:
+                host_id = self.host_map.get(host, None)
+                if host_id not in existing_host_ids:
+                    host_id = None
                 host_stats = {}
                 for stat in ('changed', 'dark', 'failures', 'ignored', 'ok', 'processed', 'rescued', 'skipped'):
                     try:
                         host_stats[stat] = self.event_data.get(stat, {}).get(host, 0)
                     except AttributeError:  # in case event_data[stat] isn't a dict.
                         pass
-                if qs.filter(name=host).exists():
-                    host_actual = qs.get(name=host)
-                    host_summary, created = job.job_host_summaries.get_or_create(host=host_actual, host_name=host_actual.name, defaults=host_stats)
-                else:
-                    host_summary, created = job.job_host_summaries.get_or_create(host_name=host, defaults=host_stats)
-
-                if not created:
-                    update_fields = []
-                    for stat, value in host_stats.items():
-                        if getattr(host_summary, stat) != value:
-                            setattr(host_summary, stat, value)
-                            update_fields.append(stat)
-                    if update_fields:
-                        host_summary.save(update_fields=update_fields)
+                summary = JobHostSummary(
+                    created=now(), modified=now(), job_id=job.id, host_id=host_id, host_name=host, **host_stats
+                )
+                summary.failed = bool(summary.dark or summary.failures)
+                summaries[(host_id, host)] = summary
+
+            JobHostSummary.objects.bulk_create(summaries.values())
+
+            # update the last_job_id and last_job_host_summary_id
+            # in single queries
+            host_mapping = dict(
+                (summary['host'], summary['id'])
```

Comment:
I don't follow something here. the keys are `summary['host']`. What is that? It looks like a `Host` object.

but later on you write `if h.id in host_mapping:`. That's looking if the id is in the keys, which it wouldn't be because the keys do not have integers, but host objects.

Better to stick to ids only when possible.

Reply:
I'm using `.values_list`, so it looks like `'host'` is just the actual ID.

I could change this to `'host_id'` instead for clarity.

```
> /awx_devel/awx/main/models/events.py(524)_update_host_summary_from_stats()
-> for h in all_hosts:
(Pdb) host_mapping.keys()
dict_keys([78, 31, 48, 82, 25, 85, 11, 98, 67, 7, 92, 59, 99, 91, 84, 60, 43, 5, 52, 24, 83, 44, 15, 3, 93, 13, 34, 79, 18, 42, 61, 27, 20, 100, 33, 69, 54, 64, 17, 66, 2, 41, 71, 89, 26, 36, 6, 40, 16, 94, 4, 46, 75, 38, 96, 68, 55, 28, 8, 76, 35, 14, 74, 95, 21, 12, 65, 81, 70, 45, 73, 53, 47, 29, 39, 9, 32, 87, 80, 19, 88, 62, 63, 72, 10, 23, 86, 77, 1, 97, 50, 51, 49, 30, 58, 37, 22, 57, 56, 90])
```